File: shared/src/commonMain/kotlin/com/vnidrop/app/feature/send/TransferShareActions.kt
package com.vnidrop.app.feature.send

import androidx.compose.runtime.Composable

enum class NfcShareAvailability { Available, Unavailable, Hidden }

interface TransferShareActions {
	val canUseNativeShare: Boolean
	val nfcAvailability: NfcShareAvailability
	fun exportInvitation(ticket: String, transferName: String, onResult: (Result<Unit>) -> Unit)
	fun shareInvitation(ticket: String, transferName: String, onResult: (Result<Unit>) -> Unit)
	fun writeInvitationToNfc(ticket: String, onResult: (Result<Unit>) -> Unit)
	fun cancelNfcWrite()
}

object UnavailableTransferShareActions : TransferShareActions {
	override val canUseNativeShare = false
	override val nfcAvailability = NfcShareAvailability.Hidden
	override fun exportInvitation(ticket: String, transferName: String, onResult: (Result<Unit>) -> Unit) =
		onResult(Result.failure(UnsupportedOperationException("Invitation export is unavailable")))
	override fun shareInvitation(ticket: String, transferName: String, onResult: (Result<Unit>) -> Unit) =
		onResult(Result.failure(UnsupportedOperationException("System sharing is unavailable")))
	override fun writeInvitationToNfc(ticket: String, onResult: (Result<Unit>) -> Unit) =
		onResult(Result.failure(UnsupportedOperationException("NFC is unavailable")))
	override fun cancelNfcWrite() = Unit
}

@Composable
expect fun rememberTransferShareActions(): TransferShareActions

internal fun invitationFileName(transferName: String): String {
	val safe = transferName.trim()
		.map { character -> if (character.isLetterOrDigit() || character in "-_. ") character else '_' }
		.joinToString("")
		.trim('.', ' ')
		.ifBlank { "VniDrop transfer" }
	return if (safe.endsWith(".vnd", ignoreCase = true)) safe else "$safe.vnd"
}

File: apple/VniDrop/Features/Send/TransferShareActions.swift
import SwiftUI

enum NfcShareAvailability { case available, unavailable, hidden }

/// Invitation delivery actions shared by the native Apple feature models.
/// Platform implementations perform export, native share, and NFC write.
@MainActor
protocol TransferShareActions: AnyObject {
	var canUseNativeShare: Bool { get }
	var nfcAvailability: NfcShareAvailability { get }

	func exportInvitation(ticket: String, transferName: String, onResult: @escaping (Result<Void, Error>) -> Void)
	func shareInvitation(ticket: String, transferName: String, onResult: @escaping (Result<Void, Error>) -> Void)
	func writeInvitationToNfc(ticket: String, onResult: @escaping (Result<Void, Error>) -> Void)
	func cancelNfcWrite()
}

/// The QR + delivery buttons for a transfer's share panel, ported from the button
/// stack in `TransferSharePanel` (`TransferDetails.kt`).
struct ShareActionsView: View {
	@Environment(\.vniColors) private var colors
	@ObservedObject var model: SendModel
	let transfer: Transfer
	let ticket: String

	@State private var actions: TransferShareActions = makePlatformShareActions()
	@State private var writingNfc = false

	var body: some View {
		VStack(spacing: 12) {
			if actions.nfcAvailability != .hidden {
				SecondaryButton(
					title: writingNfc ? String(localized: L10n.Transfer.nfcWaiting) : String(localized: L10n.Button.writeNfc),
					action: {
						writingNfc = true
						actions.writeInvitationToNfc(ticket: ticket) { result in
							writingNfc = false
							model.onInvitationResult(.nfc, result)
						}
					},
					enabled: actions.nfcAvailability == .available && !writingNfc
				)
				if actions.nfcAvailability == .unavailable {
					Text(String(localized: L10n.Transfer.nfcUnavailable))
						.font(VniType.bodySmall).foregroundStyle(colors.foregroundLighter)
				}
			}
			SecondaryButton(title: String(localized: L10n.Button.downloadInvitation), action: {
				actions.exportInvitation(ticket: ticket, transferName: transfer.transferName ?? "") {
					model.onInvitationResult(.export, $0)
				}
			})
			PrimaryButton(title: String(localized: L10n.Button.nativeShare), action: {
				actions.shareInvitation(ticket: ticket, transferName: transfer.transferName ?? "") {
					model.onInvitationResult(.share, $0)
				}
			}, enabled: actions.canUseNativeShare)
		}
		.onDisappear { actions.cancelNfcWrite() }
	}
}
